File: ca341/assignment1/compare.py
class phonebookEntry:
    def __init__(self, name, number, address):
        """Constructor for phonebookEntry class

        Args:
            name (string): name for the phonebook entry
            number (integer): number for the phonebook entry
            address (string): address for the phonebook entry
        """
        self.left = None
        self.right = None
        self.name = name
        self.number = number
        self.address = address

    def insert(self, name, number, address, sortType="name"):
        """Inserts a new phonebook entry into the phonebook, sorted by either name or number

        Args:
            name (string): name to insert into the phonebook
            number (integer): number to insert into the phonebook
            address (string): address to insert into the phonebook
            sortType (string, optional): The type of sorting to use. Defaults to "name".
        """
        if sortType == "name":
            compareOn = self.name
            compareWith = name
        elif sortType == "number":
            compareOn = self.number
            compareWith = number

        if compareOn:
            if compareWith < compareOn:
                if self.left is None:
                    self.left = phonebookEntry(name, number, address)
                else:
                    self.left.insert(name, number, address, sortType)
            elif compareWith > compareOn:
                if self.right is None:
                    self.right = phonebookEntry(name, number, address)
                else:
                    self.right.insert(name, number, address, sortType)

    def printPhonebook(self):
        """Prints the phonebook"""
        if self.left:
            self.left.printPhonebook()
        print(f"Name: {self.name}, Phone Number: {self.number}, Address: {self.address}")
        if self.right:
            self.right.printPhonebook()

    def search(self, searchFor, sortType="name"):
        """Searches for either a name and number in the phonebook and returns the full entry

        Args:
            searchFor (string): name or number to search for in the phonebook
            sortType (string, optional): The type of sorting to use. Defaults to "name".
        Returns:
            phonebookEntry: The phonebook entry with the name or the number searched for is in
        """
        if sortType == "name":
            compareOn = self.name
        elif sortType == "number":
            compareOn = self.number


        if searchFor < compareOn and self.left is not None:
            return self.left.search(searchFor, sortType)
        elif searchFor > compareOn and self.right is not None:
            return self.right.search(searchFor, sortType)
        elif searchFor == compareOn:
            return self
        else:
            return None

    def findMin(self):
        """Finds the minimum value in the given phonebook
            Used as a utility function for the delete functions

        Returns:
            list: A list containing the name and number of the minimum value
        """
        if self.left is None:
            return self
        else:
            return self.left.findMin()

    def delete(self, deleteItem, sortType="name"):
        """Deletes an entry from the phonebook based on either the name or the number

        Args:
            deleteItem (string): name or number to delete from the phonebook
            sortType (str, optional): The type of sorting to use. Defaults to "name".

        Returns:
            phonebookEntry: The phonebook with the entry deleted
        """
        if sortType == "name":
            compareOn = self.name
        elif sortType == "number":
            compareOn = self.number

        if deleteItem == usrName or deleteItem == usrNumber:
            return
        if deleteItem < compareOn:
            if self.left:
                self.left = self.left.delete(deleteItem, sortType)
        elif deleteItem > compareOn:
            if self.right:
                self.right = self.right.delete(deleteItem, sortType)
        else:
            if self.left is None and self.right is None:
                return None
            if self.left is None:
                return self.right
            if self.right is None:
                return self.left

            middleVal = self.right.findMin()
            self.name = middleVal.name
            self.number = middleVal.number
            self.address = middleVal.address
            if sortType == "name":
                self.right = self.right.delete(middleVal.name, sortType)
            elif sortType == "number":
                self.right = self.right.delete(middleVal.number, sortType)
        return self

def validateNumber(number):
    """Utility function to validate the number entered by the user

    Args:
        number (string): number entered by the user

    Returns:
        integer: The number entered by the user covnerted to an integer
    """
    while True:
        try:
            number = int(number)
            break
        except ValueError:
            print("Invalid number. Please try again.")
            number = input("Enter the number again: ")
    return number

def getUserChoice():
    """Prompts the user for input

    Returns:
        integer: The user's choice
    """
    print()
    print("1. Add a new contact")
    print("2. Delete a contact")
    print("3. Search for a contact (by name)")
    print("4. Search for a contact (by number)")
    print("5. Print all contacts")
    print("6. Exit")
    print()
    userInput = validateNumber(int(input("Enter a number: ")))
    print()
    return userInput

def userPrompt():
    """Prompts the user for input and calls the appropriate function based on the user's choice"""
    userChoice = getUserChoice()
    try:
        while userChoice != 6:
            if userChoice == 1:
                # add a new contact
                name = input("Enter a name: ")
                if root.search(name, "name") is not None:
                    print("Name already exists in the phonebook")
                    userChoice = getUserChoice()
                    continue

                number = validateNumber(int(input("Enter a phone number: ")))
                if numberRoot.search(number, "number") is not None:
                    print("Name already exists in the phonebook")
                    userChoice = getUserChoice()
                    continue
                address = input("Enter an address: ")

                root.insert(name, number, address, "name")
                numberRoot.insert(name, number, address, "number")

                print(f"\n{name} added successfully.")
                userChoice = int(getUserChoice())

            elif userChoice == 2:
                # delete a contact
                nameToDelete = input("Enter a name to delete: ")
                delete = root.search(nameToDelete, "name")
                if delete is None:
                    print("Contact not found.")

                elif delete.name == usrName:
                    print("You can't delete yourself!")
                else:
                    numberToDelete = delete.number

                    root.delete(nameToDelete, "name")
                    numberRoot.delete(numberToDelete, "number")

                    print(f"\n{name} deleted successfully.")
                userChoice = int(getUserChoice())

            elif userChoice == 3:
                # search for a contact by name
                search = input("Enter name to search for: ")
                searchResult = root.search(search, "name")

                if searchResult is not None:
                    print(f"Name: {searchResult.name}\nNumber: {searchResult.number}\nAddress: {searchResult.address}")
                else:
                    print(f"{search} Not Found")

                userChoice = int(getUserChoice())

            elif userChoice == 4:
                # search for a contact by number
                search = validateNumber(int(input("Enter number to search for: ")))
                searchResult = numberRoot.search(search, "number")

                if searchResult is not None:
                    print(f"Name: {searchResult.name}\nNumber: {searchResult.number}\nAddress: {searchResult.address}")
                else:
                    print(f"{search} Not Found")

                userChoice = int(getUserChoice())

            elif userChoice == 5:
                # print all contacts
                root.printPhonebook()
                userChoice = int(getUserChoice())

            else:
                # invalid input
                print("\nInvalid input\n")
                userChoice = int(getUserChoice())

    # if the user enters a non-integer value for the menu choice this will catch it and prompt the user to try again
    except ValueError:
        print("Invalid input")
        userPrompt()


usrName = input("Enter your name: ")
usrNumber = validateNumber(int(input("Enter your number: ")))
usrAddress = input("Enter your address: ")
root = phonebookEntry(usrName, usrNumber, usrAddress)
numberRoot = phonebookEntry(usrName, usrNumber, usrAddress)


def main():
    print(f"Welcome your phonebook, {usrName}!")

    userPrompt()

if __name__ == "__main__":
    main()

File: ca341/assignment1/compare.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_STRING_SIZE 100

struct phonebookEntry {
    char name[MAX_STRING_SIZE];
    char address[MAX_STRING_SIZE];
    int number;
    struct phonebookEntry * left;
    struct phonebookEntry * right;
};
typedef struct phonebookEntry phonebookEntry;

phonebookEntry * createEntry(char * name, char * address, int number);
phonebookEntry * insertEntry(phonebookEntry * root, char * name, char * address, int number, char * sortType);
phonebookEntry * searchEntry(phonebookEntry * root, char * name, int number);
phonebookEntry * deleteEntry(phonebookEntry * root, char * name, int number);
void printPhonebook(phonebookEntry * root);
void userPrompt(phonebookEntry * root, phonebookEntry * numberRoot);
int getNumberInput();
int getUserChoice();

int main(){
    phonebookEntry * root = NULL;
    phonebookEntry * numberRoot = NULL;

    printf("Welcome to your phonebook!\n\n");

    userPrompt(root, numberRoot);

    return 0;
}

// Creates a new phonebook entry and returns it
phonebookEntry * createEntry(char * name, char * address, int number) {
    // Allocate memory for the new entry and then copy the values into it
    phonebookEntry * newEntry = malloc(sizeof(phonebookEntry));
    strcpy(newEntry->name, name);
    strcpy(newEntry->address, address);
    newEntry->number = number;
    newEntry->left = NULL;
    newEntry->right = NULL;
    return newEntry;
}

// Inserts a new entry into the phonebook, that is sorted by name or number, in its correct position
phonebookEntry * insertEntry(phonebookEntry * root, char * name, char * address, int number, char * sortType) {

    // If the root is null, then we have reached the end of the phonebook and can insert the new entry here
    if (root == NULL) {
        return createEntry(name, address, number);
    }

    if (strcmp(sortType, "name") == 0) {
    // If the name is less than the root's name, then we need to insert the new entry to the left
    // otherwise we need to insert it to the right
        if (strcmp(name, root->name) < 0) {
            root->left = insertEntry(root->left, name, address, number, "name");
        }
        else if (strcmp(name, root->name) > 0) {
            root->right = insertEntry(root->right, name, address, number, "name");
        }
    }
    else if (strcmp(sortType, "number") == 0) {
    // If the number is less than the root's name, then we need to insert the new entry to the left
    // otherwise we need to insert it to the right
        if (number < root->number) {
            root->left = insertEntry(root->left, name, address, number, "number");
        }
        else if (number > root->number) {
            root->right = insertEntry(root->right, name, address, number, "number");
        }
    }

    return root;
}

// Searches for a given name or number in the phonebook, and returns the entry if found
phonebookEntry * searchEntry(phonebookEntry * root, char * name, int number) {
    // If the root is null, then we have reached the end of the phonebook and the entry was not found
    if (root == NULL) {
        return NULL;
    }
    if (strcmp(name, "") != 0) {
        // If the name is less than the root's name, then we need to search the left subtree
        // If the name is greater than the root's name, then we need to search the right subtree
        // Otherwise we have found the entry and can return it
        if (strcmp(name, root->name) < 0) {
            return searchEntry(root->left, name, 0);
        }
        else if (strcmp(name, root->name) > 0) {
            return searchEntry(root->right, name, 0);
        }
        else {
            return root;
        }
    }
    else if (number != 0) {
        // If the number is less than the root's number, then we need to search the left subtree
        // If the number is greater than the root's number, then we need to search the right subtree
        // Otherwise we have found the entry and can return it
        if (number < root->number) {
            return searchEntry(root->left, "", number);
        }
        else if (number > root->number) {
            return searchEntry(root->right, "", number);
        }
        else {
            return root;
        }
    }

}

// Deletes a given entry from the phonebook, that is sorted by name or number, and returns the new root
// https://www.geeksforgeeks.org/deletion-in-binary-search-tree/
phonebookEntry * deleteEntry(phonebookEntry * root, char * name, int number) {
    // If the root is null, then we have reached the end of the phonebook and the entry was not found and could not be deleted
    if (root == NULL) {
        return NULL;
    }

    // In both cases, we need to search for the entry to be deleted, if the value is less than the root's value, then we need to search the left subtree, otherwise we need to search the right subtree, if we find the entry, then we can delete it

    if (strcmp(name, "") != 0) {
        // If the name is less than the root's name, then the left subtree will be searched for the entry
        if (strcmp(name, root->name) < 0) {
            root->left = deleteEntry(root->left, name, 0);
        }
        // If the name is greater than the root's name, then the right subtree will be searched for the entry
        else if (strcmp(name, root->name) > 0) {
            root->right = deleteEntry(root->right, name, 0);
        }
        // If the name is equal to the root's name, then this is the entry to be deleted
        else {
            if (root->left == NULL) {
                phonebookEntry * temp = root->right; // Store the subtree in a temporary variable
                free(root); // Free the memory used by the root
                return temp; // Return the subtree as the new root
            }
            else if (root->right == NULL) {
                phonebookEntry * temp = root->left;
                free(root);
                return temp;
            }

            // If the root has two children, then the right subtree must be searched for the smallest entry
            phonebookEntry * temp = root->right; // Store the right subtree in a temporary variable
            // Loop down the left subtree to find the smallest entry
            // The final value of temp will be the smallest entry in the right subtree and will replace the root
            while (temp->left != NULL) {
                temp = temp->left;
            }
            strcpy(root->name, temp->name);
            strcpy(root->address, temp->address);
            root->number = temp->number;
            // To not have a duplicate entry in the phonebook, the smallest entry in the right subtree is deleted
            root->right = deleteEntry(root->right, temp->name, 0);
        }
    }
    else if (number != 0) {
        if (number < root->number) {
            root->left = deleteEntry(root->left, "", number);
        }
        else if (number > root->number) {
            root->right = deleteEntry(root->right, "", number);
        }
        else {
            if (root->left == NULL) {
                phonebookEntry * temp = root->right;
                free(root);
                return temp;
            }
            else if (root->right == NULL) {
                phonebookEntry * temp = root->left;
                free(root);
                return temp;
            }

            phonebookEntry * temp = root->right;
            while (temp->left != NULL) {
                temp = temp->left;
            }
            strcpy(root->name, temp->name);
            strcpy(root->address, temp->address);
            root->number = temp->number;
            root->right = deleteEntry(root->right, "", temp->number);
        }
    }
    return root;
}

// Prints the phonebook in alphabetical order
void printPhonebook(phonebookEntry * root) {
    // First the left subtree is printed, then the root, then the right subtree
    if (root == NULL) {
        return;
    }
    printPhonebook(root->left);
    printf("Name: %s, Address: %s, Phone Number: %d\n", root->name, root->address, root->number);
    printPhonebook(root->right);
}

// Gets the user's input and parses it to an integer
int getNumberInput() {
    int number;
    char input[MAX_STRING_SIZE];
    fgets(input, MAX_STRING_SIZE, stdin);

    // https://stackoverflow.com/a/1248017 - Remove newline character at the end of the string and replace it with a null terminator
    if ((strlen(input) > 0) && (input[strlen(input) - 1] == '\n')) {
        input[strlen (input) - 1] = '\0';
    }

    // strtol is used to convert string to int
    number = strtol(input, NULL, 10);
    if (number < 1) {
        printf("Invalid input.\nInput can only be positive, less than 11 digits long and no larger than 2147483647\nEnter number: ");
        return getNumberInput();
    }

    return number;
}

// Prompts the user for a choice and returns it
int getUserChoice() {
    int choice;
    printf("\n");
    printf("1. Add a new contact\n");
    printf("2. Delete a contact\n");
    printf("3. Search for a contact (by name)\n");
    printf("4. Search for a contact (by number)\n");
    printf("5. Print the phonebook\n");
    printf("6. Exit\n");
    printf("\nEnter a number: ");

    choice = getNumberInput();
    return choice;
}

// Takes the user's choice and performs the appropriate action
void userPrompt(phonebookEntry * root, phonebookEntry * numberRoot) {
    int choice = getUserChoice();
    printf("\n");
    char name[MAX_STRING_SIZE];
    char address[MAX_STRING_SIZE];
    int number;
    char input[MAX_STRING_SIZE];
    phonebookEntry * temp = malloc(sizeof(phonebookEntry));
    phonebookEntry * searchResult = malloc(sizeof(phonebookEntry));
    // Switch statement to perform the appropriate action based on the user's choice
    // Will loop until the user chooses to exit by inputting 6
    while (choice != 6) {
        switch (choice) {
            case 1:
                // Adding a new contact to the phonebook

                printf("Enter a name: ");
                fgets(name, MAX_STRING_SIZE, stdin);
                if ((strlen(name) > 0) && (name[strlen(name) - 1] == '\n')) {
                    name[strlen (name) - 1] = '\0';
                }
                // check if the name already exists in the phonebook
                searchResult = searchEntry(root, name, 0);
                if (searchResult != NULL) {
                    printf("A contact with that name already exists.\n");
                    break;
                }

                printf("Enter a phone number: ");
                number = getNumberInput();
                if (number == 0) {
                    printf("Invalid phone number entered. Please try again.\n");
                    break;
                }
                // check if the number already exists in the phonebook
                searchResult = searchEntry(numberRoot, "", number);
                if (searchResult != NULL) {
                    printf("A contact with that number already exists.\n");
                    break;
                }

                printf("Enter an address: ");
                fgets(address, MAX_STRING_SIZE, stdin);
                if ((strlen(address) > 0) && (address[strlen(address) - 1] == '\n')) {
                    address[strlen (address) - 1] = '\0';
                }

                root = insertEntry(root, name, address, number, "name");
                numberRoot = insertEntry(numberRoot, name, address, number, "number");

                printf("\n%s added successfully.\n", name);
                break;

            case 2:
                // Deleting a contact from the phonebook

                printf("Enter a name to delete: ");
                fgets(name, MAX_STRING_SIZE, stdin);
                if ((strlen(name) > 0) && (name[strlen(name) - 1] == '\n')) {
                    name[strlen (name) - 1] = '\0';
                }

                // The entry needs to be deleted from both the name and number sorted phonebooks
                // To be found in the number sorted phonebook, the number needs to be found first from the name sorted phonebook
                temp = searchEntry(root, name, 0);
                if (temp == NULL) {
                    printf("Contact not found.\n");
                    break;
                }
                int tempNumber = temp->number;
                root = deleteEntry(root, name, 0);
                numberRoot = deleteEntry(numberRoot, "", tempNumber);
                printf("\n%s deleted successfully\n", name);
                temp = NULL;
                break;

            case 3:
                // Searching for a contact by name

                printf("Enter name to search for: ");
                fgets(name, MAX_STRING_SIZE, stdin);
                if ((strlen(name) > 0) && (name[strlen(name) - 1] == '\n')) {
                    name[strlen (name) - 1] = '\0';
                }

                phonebookEntry * searchResult = searchEntry(root, name, 0);
                if (searchResult != NULL) {
                    printf("Name: %s\n", searchResult->name);
                    printf("Address: %s\n", searchResult->address);
                    printf("Phone number: %d\n", searchResult->number);
                }
                else {
                    printf("Contact not found\n");
                }
                break;

            case 4:
                // Searching for a contact by number

                printf("Enter number to search for: ");
                number = getNumberInput();
                phonebookEntry * searchNumberResult = searchEntry(numberRoot, "", number);
                if (searchNumberResult != NULL) {
                    printf("Name: %s\n", searchNumberResult->name);
                    printf("Address: %s\n", searchNumberResult->address);
                    printf("Phone number: %d\n", searchNumberResult->number);
                }
                else {
                    printf("Contact not found\n");
                }
                break;

                case 5:
                    // Printing the phonebook
                    printPhonebook(root);
                    break;

            default:
                printf("Please enter a valid option\n\n");
                break;
        }
        choice = getUserChoice();
        printf("\n");
    }
    free(temp);
}